### PR TITLE
Correctly remove orphaned non-tree subgraphs

### PR DESCRIPTION
**Pull request by mischnic**
# ↪️ Pull Request

`removeNode(1)`  first removes 0-1, then 1-2 (because 2 is an orphan), then 2-3 (because 2 is an orphan after 0-1 was removed). The the right branch is visited and 2-3 is removed again which fails

undefined

I think it makes more sense to first remove the outgoing edges and then the incoming edges. This way, the subgraph is removed bottom up and a nodes becomes only orphaned once the last parent of that node is removed. So there will always be exactly one anchestor node that is responsible for removing it and no double deletion as before.

## 💻 Examples

The asset graph looked like this, changing some target option in package.json would remove the subgraph and run into a `Edge from x to y not found`  error.

### CODE REVIEW

File: packages/core/graph/test/Graph.test.js [hunk 1]
```diff
@@ -317,4 +317,24 @@ describe('Graph', () => {
 
     assert.deepEqual(visited, [nodeA, nodeB, nodeD]);
   });
+
+  it('foo', () => {
```
Comment (wbinnssmith) [inline packages/core/graph/test/Graph.test.js]: placeholder? 😁

Comment (mischnic) [reply packages/core/graph/test/Graph.test.js]:
```suggestion
  it('correctly removes non-tree subgraphs', () => {
```

**Review by wbinnssmith — APPROVED** [review]

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 9.78s | +431.00ms |
| Cached | 475.00ms | -11.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>AtlasKit Editor 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.55s | -84.00ms |
| Cached | 300.00ms | +20.00ms ⚠️ |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/Three.js | 579.69kb | +0.00b | 5.40s | +728.00ms ⚠️ |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/329120273400857170)**

**Comment by mischnic** [comment]
Now this tests fails where b is removed and d is deleted twice
undefined

I wonder if we should just make the assertions  in the methods less strict.

**Comment by mischnic** [comment]
@lettertwo @thebriando What do you think about removing some of these assertions in `removeNode` and `removeEdge`? I think finding an ordering/algorithm that doesn't have "double deletes" will be more effort (certainly to write it, and possibly also computationally) that just ignoring these additional deletions.

**Comment by lettertwo** [comment]
> What do you think about removing some of these assertions in `removeNode` and `removeEdge`?

I was wondering this, too. I think we already do silent no-ops for the inverse operations (`addNode` and `addEdge`), so symmetrically, it would make sense to me. I don't currently know of a reason why would need to assert that a node or edge exists before removing it. I imagine any case that hinges on the success or failure of removal could be satisfied by returning a success boolean instead.

**Comment by thebriando** [comment]
> What do you think about removing some of these assertions in removeNode and removeEdge?

Yep that makes sense to me as well.

**Comment by lettertwo** [comment]
I can't find a place where the result of `removeEdge` is handled. The only places I see where `removeNode` results are handled are in [these](https://github.com/parcel-bundler/parcel/blob/84847ed76aca867ce6f5d11f3d783672dc8088fe/packages/core/graph/test/Graph.test.js#L25-L27) [two](https://github.com/parcel-bundler/parcel/blob/84847ed76aca867ce6f5d11f3d783672dc8088fe/packages/core/graph/test/Graph.test.js#L264-L270) tests

**Comment by lettertwo** [comment]
I did find this old PR: https://github.com/parcel-bundler/parcel/pull/3562, but I'm not sure if the context around it is still relevant. /cc @wbinnssmith

**Review by thebriando — APPROVED** [review]

**Review by lettertwo — APPROVED** [review]

**Pull request merged**